com.unity.render-pipelines.high-definition/Editor/Material/Nature/SpeedTree8MaterialUpgrader.cs

```diff
@@ -0,0 +1,146 @@
+            // property values but does have the correct keywords.
+            // In both cases, a keyword is only ever set if it is correct.
+            // So in order to sync keywords with exposed properties, we check first if a relevant keyword is set.
+            // Only if it is not set do we check the property to get the value.
+            // For wind in particular, we only ever pass in non-negative windInt when we intend to set it to that value.
+            // So in summary:
+            // Priority order of property-controlled keywords that require syncing:
+            // EFFECT_BILLBOARD: enabled keyword > property value
+            // _WINDQUALITY*: valid windInt > enabled keyword > property value
+
+            if (material != null)
+            {
+                if (material.shader.name.Equals("HDRP/Nature/SpeedTree8"))
+                {
+                    material.EnableKeyword("_ALPHATEST_ON");
+                    material.EnableKeyword("ENABLE_WIND");
+                    // Assumes all ST8 materials are double-sided.
+                    material.EnableKeyword("_DOUBLESIDED_ON");
+
+                    if (!WindIntValid(windInt))
+                    {
+                        windInt = GetWindQualityFromKeywords(material.shaderKeywords);
+                        if (!WindIntValid(windInt))
+                            windInt = material.HasFloat("_WindQuality") ? (int)material.GetFloat("_WindQuality") : 0;
+                        if (!WindIntValid(windInt))
+                            windInt = 0;
+                    }
+
+                    material.EnableKeyword(WindQualityString[windInt]);
+                    material.SetFloat("_WindQuality", windInt); // A legacy float used in native code to apply wind data
+                    material.SetFloat("_WINDQUALITY", windInt); // The actual name of the keyword enum for the shadergraph
+
+                    // Shadergraph sets cull mode from _CullMode, not _TwoSided (which is what ST uses)
+                    if (material.name.Contains("Billboard"))
+                    {
+                        bool billboardEnabled = material.IsKeywordEnabled("EFFECT_BILLBOARD")
+                            || material.GetFloat("EFFECT_BILLBOARD") != 0;
+                        if (billboardEnabled)
+                        {
+                            material.EnableKeyword("EFFECT_BILLBOARD");
+                            material.SetFloat("EFFECT_BILLBOARD", 1);
+                        }
+                        else
+                        {
+                            material.DisableKeyword("EFFECT_BILLBOARD");
+                            material.SetFloat("EFFECT_BILLBOARD", 0);
+                        }
+                        // Billboards have double sided mode = none.
+                        material.SetVector("_DoubleSidedConstants", kDoubleSidedNone);
```

Comment:
Are billboard double sided ? mode = none means that normals on the backface will be set equal to normals on the front face

Reply:
Billboards get backfaces culled, so they have double sided mode disabled. Or is there another mode I should use?

Reply:
If backfaces are culled the mode doesn't have imoprtance, but I think then the keyword _DOUBLESIDED_ON should be disabled (currently it's always enabled at the beginning of the function)

Reply:
That makes sense!